### DISCUSSION TITLE
WeAct Blackpill STM32F411 pin.irq usage

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#16247] [STM32 / Pyboard] [answered]
After many attempts to get 3 interrupt sources working reliably I was left wondering ... whether or not there was some obvious reason,

The code for one sensor:
```py
def gate_interrupt(pin):
    """Gate sensor interrupt."""
    global gate_flag
    global cars_coming_in
    global cars_going_out
    global cars_in_at_gate
    global cars_out_at_gate

    gate_sensor.irq(None, Pin.IRQ_FALLING)

    gate_flag = True

    time.sleep_ms(DEBOUNCE)

    if cars_coming_in is True:
        cars_in_at_gate += 1
    elif cars_going_out is True:
        cars_out_at_gate += 1

    gate_sensor.irq(gate_interrupt, Pin.IRQ_FALLING)
```
I read somewhere that `gate_sensor.irq(None, Pin.IRQ_FALLING)` would clear the IRQ buffer of any pending interrupt and I guess I hoped that no more interrupts would be saved until you hit `gate_sensor.irq(gate_interrupt, Pin.IRQ_FALLING)`

This not seem to be the case.  I am aware that ISRs should be short and I am even putting `disable_irq` and `enable_irq` around critical sections in the main code.

I think pending interrupts are stacking-up somewhere. Any suggestions for my next attempt would be appreciated

**Comment by rkompass** [comment]
AFAIK `gate_sensor.irq()` is no part of standard MP.
So I cannot understand how the code example is about to work.

> I read somewhere that gate_sensor.irq(None, Pin.IRQ_FALLING) would clear the IRQ buffer of any pending interrupt ...

Can you provide a link to the relevant docs/descriptions?

**Reply by scruss** [reply]
If `gate_sensor` is a Pin object, then it may refer to [Pin.irq()](https://docs.micropython.org/en/v1.24.0/library/machine.Pin.html#machine.Pin.irq)

**Reply by davefes** [reply]
I should have been more explicit ... at the top of the script there is:
`gate_sensor.irq(trigger=Pin.IRQ_FALLING, handler=gate_interrupt)`

**Reply by davefes** [reply]
@rkompass 

https://www.i-programmer.info/programming/148-hardware/17088-esp32-in-micropython-interrupts.html?start=1

> You can clear the queue of interrupts waiting to be handled by disabling interrupts within the interrupt handler and then turning them back on again:

Perhaps the STM32 interrupt system is different.

**Comment by shariltumin** [comment]
Hi Dave, you can have a look at the code below, 

```py
import machine, time
from machine import Pin
from _thread import start_new_thread as task

def pin_cb(p):
    global t0, t1
    wait = W - time.ticks_diff(t1, t0)
    if wait > 0:
       print(f'Event happend, do nothing. Wait {wait} ms')
       t1 = time.ticks_ms()
    else:
       print("hello", p, time.ticks_us())
       t0 = time.ticks_ms()

touch = Pin('A0', Pin.OUT, value=0)
touch.irq(trigger=Pin.IRQ_FALLING, handler=pin_cb)

t0 = t1 = 0; more = True; W = 5000

def loop():
   while more:
       touch(1)
       touch(0) # trigger
       print('Sleep in loop'); time.sleep_ms(800)

task(loop, ()) # start loop

time.sleep(30) # keep all alive for 60 sec
print('Disable handler')
touch.irq(trigger=Pin.IRQ_FALLING, handler=None) # disable handler
time.sleep(5)
more = False
print('Bye')
```

hopefully it helps.

**Reply by davefes** [reply]
@shariltumin, thanks for the example.

I have spent the last few days working on an `asyncio` solution, without interrupts, which seems to be working better.

Threading is a totally new topic for me, but I will have a go at your suggestion.

**Comment by rkompass** [comment, marked as answer]
@davefes 

> I think pending interrupts are stacking-up somewhere. Any suggestions for my next attempt would be appreciated

I had this problem when I noticed that error message:  `RuntimeError: schedule queue full`.

I asked for help in [this discussion thread](https://github.com/orgs/micropython/discussions/13187) and recommend reading it.

**Comment by davefes** [comment]
@rkompass informative reading.  I now understand that this problem requires serious work.  Thanks for the suggestions.
